docs/guide/programming-tasks.md

```diff
@@ -0,0 +1,288 @@
+2. Starlark script can be executed instantly on any platform. There is Starlark interpreter written in Go which integrates nicely with Cirrus CLI / Cirrus CI infrastructure.
+3. Starlark has built-in functionality for loading external modules which is ideal for config sharing. See [module loading](#module-loading) for details.
+
+## Writing Starlark scripts
+
+Let's start with a trivial `.cirrus.star` example like this:
+
+```python
+def main():
+    return [
+        {
+            "container": {
+                "image": "debian:latest",
+            },
+            "script": "make",
+        },
+    ]
+```
+
+`main` simply returns a list of task objects which will be serialized into YAML presentation like this:
+
+```yaml
+task:
+    container:
+      image: debian:latest
+    script: make
+```
+
+Then the generated YAML is appended to `.cirrus.yml` (if any) before passing the combined config into the final YAML parser.
+
+With Starlark, it's possible to generate parts of the configuration dynamically based on some external conditions: by [making an HTTP request](#http) to check the previous build status
+or by [parsing files inside the repository](#fs) to pick up some common settings (for example, parse `package.json` to see if it contains `lint` script and generate a linting task).
+
+Even more importantly: with the [module loading](#module-loading) you can re-use other people's code to avoid wasting time on things written from scratch.
+For example, there are official [task helpers](https://github.com/cirrus-modules/helpers) available that reduce the boilerplate when generating tasks:
+
+```python
+load("github.com/cirrus-modules/helpers", "task", "container", "script")
+
+def main(ctx):
+  return [
+    task(
+      instance=container("debian:latest"),
+      instructions=[script("make")]
+    ),
+  ]
+```
+
+### Entrypoints
+
+Different events will trigger execution of different top-level functions in the `.cirrus.star`. These functions reserve certain names
+and will be called with different arguments depending on the event which triggered the execution.
+
+#### `main`
+
+`main` is called once a Cirrus CI build is triggered in order to generate a list of tasks to execute within that particular build.
+
+#### Hooks
+
+It's also possible to execute Starlark scripts on updates to the current build or any of the tasks within the build.
+Think of it as [WebHooks](../api.md#webhooks) running within Cirrus that doesn't require any infrastructure on your end.
+
+Expected names of Starlark Hook functions in `.cirrus.star` are `on_build_<STATUS>` or `on_task_<STATUS>` respectively.
+These functions should expect a single context argument passed by Cirrus Cloud. At the moment hook's context only contains
+a single field `payload` containing the [same payload as a webhook](../api.md#webhooks).
+
+One caveat of Starlark Hooks execution is `CIRRUS_TOKEN` [environment variable](#env) that contains a token to access [Cirrus API](../api.md).
```

Comment:
There are both Starlark Hooks and Cirrus Hooks occurrences across the document, perhaps we should only pick one?

Reply:
Fixed in 9ffc826ac963948e94024fc64d696a2eeaad843c